New Resource: `azurerm_redhat_openshift_cluster` (#24375)
hashicorp/terraform-provider-azurerm: website/docs/r/redhat_openshift_cluster.html.markdown

---
subcategory: "Red Hat OpenShift"
layout: "azurerm"
page_title: "Azure Resource Manager: azurerm_redhat_openshift_cluster"
description: |-
  Manages fully managed Azure Red Hat OpenShift Cluster (also known as ARO)
---

# azurerm_redhatopenshift_cluster

Manages a fully managed Azure Red Hat OpenShift Cluster (also known as ARO).

~> **Note:** All arguments including the client secret will be stored in the raw state as plain-text. [Read more about sensitive data in state](/docs/state/sensitive-data.html).

## Example Usage

```hcl

data "azurerm_client_config" "example" {}

data "azuread_client_config" "example" {}

resource "azuread_application" "example" {
  display_name = "example-aro"
}

resource "azuread_service_principal" "example" {
  application_id = azuread_application.example.application_id
}

resource "azuread_service_principal_password" "example" {
  service_principal_id = azuread_service_principal.example.object_id
}

data "azuread_service_principal" "redhatopenshift" {

>   // This is the Azure Red Hat OpenShift RP service principal id, do NOT use resource to reference it

Reviewer: ```suggestion
  // This is the Azure Red Hat OpenShift RP service principal id, do NOT delete it
```

  depends_on = [
    "azurerm_role_assignment.role_network1",
    "azurerm_role_assignment.role_network2",
  ]
}

output "console_url" {
  value = azurerm_redhatopenshift_cluster.example.console_url
}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required) The name of the Azure Red Hat OpenShift Cluster to create. Changing this forces a new resource to be created.

* `location` - (Required) The location where the Azure Red Hat OpenShift Cluster should be created. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) Specifies the Resource Group where the Azure Red Hat OpenShift Cluster should exist. Changing this forces a new resource to be created.

* `service_principal` - (Required) A `service_principal` block as defined below.

* `main_profile` - (Required) A `main_profile` block as defined below.

* `worker_profile` - (Required) A `worker_profile` block as defined below.

* `cluster_profile` - (Required) A `cluster_profile` block as defined below.

* `api_server_profile` - (Required) An `api_server_profile` block as defined below.

* `ingress_profile` - (Required) An `ingress_profile` block as defined below.

> * `network_profile` - (Optional) A `network_profile` block as defined below.

Reviewer: ```suggestion
* `network_profile` - (Required) A `network_profile` block as defined below.
```

* `tags` - (Optional) A mapping of tags to assign to the resource.

---

A `service_principal` block supports the following:

> * `client_id` - (Required) The Client ID for the Service Principal. Changing this forces a new resource to be created.

Reviewer: is `client_id` forcenew?

> * `client_secret` - (Required) The Client Secret for the Service Principal. Changing this forces a new resource to be created.

Reviewer: is `client_secret` forcenew?

~> **Note:** Currently a service principal cannot be associated with more than one ARO clusters on the Azure subscription.

---

A `main_profile` block supports the following:

* `subnet_id` - (Required) The ID of the subnet where main nodes will be hosted. Changing this forces a new resource to be created.

* `vm_size` - (Required) The size of the Virtual Machines for the main nodes. Changing this forces a new resource to be created.

* `encryption_at_host_enabled` - (Optional) Whether main virtual machines are encrypted at host. Defaults to `false`. Changing this forces a new resource to be created.

> **NOTE:** `encryption_at_host_enabled` is only available for certain VM sizes and the `EncryptionAtHost` feature must be enabled for your subscription. Please see the [Azure documentation](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-host-based-encryption-portal?tabs=azure-powershell) for more information.

Reviewer: ```suggestion
**NOTE:** `encryption_at_host_enabled` is only available for certain VM sizes and the `EncryptionAtHost` feature must be enabled for your subscription. Please see the [Azure documentation](https://learn.microsoft.com/azure/virtual-machines/disks-enable-host-based-encryption-portal?tabs=azure-powershell) for more information.
```

* `disk_encryption_set_id` - (Optional) The resource ID of an associated disk encryption set. Changing this forces a new resource to be created.

---

A `worker_profile` block supports the following:

* `subnet_id` - (Required) The ID of the subnet where worker nodes will be hosted. Changing this forces a new resource to be created.

* `vm_size` - (Required) The size of the Virtual Machines for the worker nodes. Changing this forces a new resource to be created.

* `disk_size_gb` - (Required) The internal OS disk size of the worker Virtual Machines in GB. Changing this forces a new resource to be created.

* `node_count` - (Required) The initial number of worker nodes which should exist in the cluster. Changing this forces a new resource to be created.

* `encryption_at_host_enabled` - (Optional) Whether worker virtual machines are encrypted at host. Defaults to `false`. Changing this forces a new resource to be created.

> **NOTE:** `encryption_at_host_enabled` is only available for certain VM sizes and the `EncryptionAtHost` feature must be enabled for your subscription. Please see the [Azure documentation](https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-host-based-encryption-portal?tabs=azure-powershell) for more information.

Reviewer: ```suggestion
**NOTE:** `encryption_at_host_enabled` is only available for certain VM sizes and the `EncryptionAtHost` feature must be enabled for your subscription. Please see the [Azure documentation](https://learn.microsoft.com/azure/virtual-machines/disks-enable-host-based-encryption-portal?tabs=azure-powershell) for more information.
```

* `domain` - (Required) The custom domain for the cluster. For more info, see [Prepare a custom domain for your cluster](https://docs.microsoft.com/azure/openshift/tutorial-create-cluster#prepare-a-custom-domain-for-your-cluster-optional). Changing this forces a new resource to be created.

* `pull_secret` - (Optional) The Red Hat pull secret for the cluster. Changing this forces a new resource to be created.

* `fips_enabled` - (Optional) Whether Federal Information Processing Standard (FIPS) validated cryptographic modules are used. Defaults to `false`. Changing this forces a new resource to be created.

---

A `network_profile` block supports the following:

* `pod_cidr` - (Required) The CIDR to use for pod IP addresses. Changing this forces a new resource to be created.

* `service_cidr` - (Required) The network range used by the OpenShift service. Changing this forces a new resource to be created.

* `outbound_type` - (Optional) The outbound (egress) routing method. Possible values are `Loadbalancer` and `UserDefinedRouting`. Defaults to `LoadBalancer`. Changing this forces a new resource to be created.

---

A `api_server_profile` block supports the following:

> * `visibility` - (Optional) Cluster API server visibility. Supported values are `Public` and `Private`. Defaults to `Public`. Changing this forces a new resource to be created.

Reviewer: ```suggestion
* `visibility` - (Required) Cluster API server visibility. Supported values are `Public` and `Private`. Defaults to `Public`. Changing this forces a new resource to be created.
```

---

A `ingress_profile` block supports the following:

> * `visibility` - (Optional) Cluster Ingress visibility. Supported values are `Public` and `Private`. Defaults to `Public`. Changing this forces a new resource to be created.

Reviewer: ```suggestion
* `visibility` - (Required) Cluster Ingress visibility. Supported values are `Public` and `Private`. Defaults to `Public`. Changing this forces a new resource to be created.
```
